### PR TITLE
feat: make `TerminatableTask` terminate itself when dropped

### PR DESCRIPTION
**Pull request by YuanYuYuan**
This PR introduces an auto termination when dropped and resolves the bug in #1054.

### CODE REVIEW

**Comment by YuanYuYuan** [comment]
@wyfo could you please review this? Thanks.

File: commons/zenoh-task/src/lib.rs [hunk 1]
```diff
@@ -168,24 +174,26 @@ impl TerminatableTask {
...
-            handle: rt.spawn(task),
+            handle: Some(rt.spawn(task)),
             token,
         }
     }
 
     /// Attempts to terminate the task.
     /// Returns true if task completed / aborted within timeout duration, false otherwise.
-    pub fn terminate(self, timeout: Duration) -> bool {
+    pub fn terminate(&mut self, timeout: Duration) -> bool {
```
Comment (wyfo) [inline commons/zenoh-task/src/lib.rs:184]: I don't think you should change the receiver type of the method

File: commons/zenoh-task/src/lib.rs [hunk 2]
```diff
@@ -168,24 +174,26 @@ impl TerminatableTask {
...
         ResolveFuture::new(async move { self.terminate_async(timeout).await }).res_sync()
     }
 
     /// Async version of [`TerminatableTask::terminate()`].
-    pub async fn terminate_async(self, timeout: Duration) -> bool {
+    pub async fn terminate_async(&mut self, timeout: Duration) -> bool {
```
Comment (wyfo) [inline commons/zenoh-task/src/lib.rs:189]: Idem about the receiver type

**Review by wyfo — COMMENTED** [review]
I see why you changed the receiver type, but I think you should not try to reuse `terminate` in `drop`.
Instead, you can use some intermediate function:
```rust
async fn terminate_inner(handle: JoinHandle<()>, timeout: Duration) -> bool {
    if tokio::time::timeout(timeout, handle).await.is_err() {
        tracing::error!("Failed to terminate the task");
        return false;
    }
    true
}
```
And use it like
```rust
fn drop(&mut self) {
    if let Some(handle) = self.handle.take() {
        ResolveFuture::new(terminate_inner(handle, std::time::Duration::from_secs(10))).res_sync();
    }
}

fn terminate(self, timeout: Duration) -> bool {
    ResolveFuture::new(terminate_inner(self.handle.take().unwrap(), timeout)).res_sync();
}

fn terminate_async(self, timeout: Duration) -> bool {
    terminate_inner(self.handle.take().unwrap(), timeout).await
}
```

**Comment by YuanYuYuan** [comment]
If I'm not mistaken, to take out the value inside an `Option` requires `&mut self`. So we need to modify `terminate` and `terminate_async` in any way. Furthermore, we discussed with @Mallets today that moving out the ownership is not necessary as the borrow checker only lives in Rust, and mixing deconstruction and undeclaration could cause some conflicts. (You can check `Undeckarable` for more details.) We could follow the common pattern like https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close. Closing a socket or receiver simply takes `&mut self` and hence `Drop` can coexist. I will create another draft PR on this topic further. 

In conclusion, I suggest splitting the termination and the deconstruction of the task. @wyfo  what do you think?

**Comment by wyfo** [comment]
> If I'm not mistaken, to take out the value inside an Option requires &mut self. So we need to modify terminate and terminate_async in any way. 

Just use `mut self` instead of `self`, and you can still use value semantic. 

> We could follow the common pattern like https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close.

This is not the same thing here, as `terminate` returns a boolean. Do you want to store this boolean to return it on following calls?

> Closing a socket or receiver simply takes &mut self and hence Drop can coexist.

There is no issue to coexist with drop and having `terminate` with `self` receiver.

**Comment by YuanYuYuan** [comment]
Hi @wyfo , thanks for you feedback.

Yes. That's what I mean we need to make some change in any way. Since I saw you wrote this in your first version, I thought you want to avoid any modification on `self`.
```rust
fn terminate(self, timeout: Duration) -> bool {
    ResolveFuture::new(terminate_inner(self.handle.take().unwrap(), timeout)).res_sync();
}
```

And I'm fine with the change of `fn terminate(mut self)`.


The remaining issue is do we want to enforce a descrtuction here? This is related to the potential conflict with `Drop`. Let me explain it.

Originally, we define our `TerminatableTask` without `Option`

```rust
pub struct TerminatableTask {
    handle: JoinHandle<()>,
    token: CancellationToken,
}
```
This forbids us from implementing `Drop` for it. The conflict comes from

```rust
async fn terminate_inner(handle: JoinHandle<()>, timeout: Duration) -> bool {
    if tokio::time::timeout(timeout, handle).await.is_err() {
        tracing::error!("Failed to terminate the task");
        return false;
    }
    true
}
impl TerminatableTask {
    pub async fn terminate_async(self, timeout: Duration) -> bool {
        terminate_inner(self.handle, timeout).await  // ERROR: cannot move out of type `TerminatableTask`, which implements the `Drop` trait
    }
}
```

Passing `self` into the function potentially trigger a deconstruction or move out itself. There are same issues in the `Undeclarable` trait.
For instance, [we implement `Undeclarable` trait for the `Subscriber`](https://github.com/eclipse-zenoh/zenoh/blob/main/zenoh/src/subscriber.rs#L803-L807), then we can't implement `Drop` for it anymore.

```rust
impl<R> Drop for Subscriber<'_, R> {
    fn drop(&mut self) {
        todo!()
    }
}
```

(You can check this does cause lots of "cannot move out ..." errors)

So that's why I need to wrap the `JoinHandle` with a `Option` to not take out the handle but the inner value.

Back to this PR, either using `&mut self` or `mut self` can compile. But we still need to think twice: do we need to embed a deconstruction in `terminate`? As the example I pasted, closing a socket multiple times is acceptable in real use cases. In this PR, a 2nd termination on a closed task is simply doing nothing. This is exactly the same with the previous behavior in zenoh library (before the tokio porting).

Moreover, it can avoid any conflict with `Drop` if we implment/modify the `TerminatableTask` as below one day (which unfortunately has already existed in the case of `Undeclarable`).

```rust
impl TerminatableTask {
    pub async fn terminate_async(mut self, timeout: Duration) -> bool {
        let Self { token, mut handle } = self;  // ERROR: cannot move out of type `TerminatableTask`, which implements the `Drop` trait
        token.cancel();
        if let Some(handle) = handle.take() {
            if tokio::time::timeout(timeout, handle).await.is_err() {
                tracing::error!("Failed to terminate the task");
                return false;
            };
        }
        true
    }
}
```

BTW, I'm little bit confused with your point "should not try to reuse terminate in drop". TBH, I don't care about the return boolean especially it would report an error once it failed. And it seems that your example don't process the boolean result in `Drop` as well. Am I missing something?

**Comment by wyfo** [comment]
I give way to your arguments, it's good to me.

**Review by Mallets — APPROVED** [review]
